packages/react-scheduler/src/ReactScheduler.js

```diff
@@ -85,12 +91,34 @@ if (!ExecutionEnvironment.canUseDOM) {
     clearTimeout(timeoutID);
   };
 } else {
-  // Always polyfill requestIdleCallback and cancelIdleCallback
+  // We keep callbacks in a queue.
+  // Calling rIC will push in a new callback at the end of the queue.
+  // When we get idle time, callbacks are removed from the front of the queue
+  // and called.
+  const pendingCallbacks: Array<CallbackConfigType> = [];
 
-  let scheduledRICCallback = null;
-  let isIdleScheduled = false;
-  let timeoutTime = -1;
+  // Number.MAX_SAFE_INTEGER is not supported in IE
+  const MAX_SAFE_INTEGER = Number.MAX_SAFE_INTEGER || 9007199254740991;
+  let callbackIdCounter = 1;
+  const getCallbackId = function(): number {
+    callbackIdCounter =
+      callbackIdCounter >= MAX_SAFE_INTEGER ? 1 : callbackIdCounter + 1;
+    return callbackIdCounter;
+  };
+
+  // When a callback is scheduled, we register it by adding it's id to this
+  // object.
+  // If the user calls 'cIC' with the id of that callback, it will be
+  // unregistered by removing the id from this object.
+  // Then we skip calling any callback which is not registered.
+  // This means cancelling is an O(1) time complexity instead of O(n).
+  const registeredCallbackIds: {[number]: boolean} = {};
```

Comment:
Use a Map?

Reply:
That was in reference to using a linked list entry instead, not if we're going to use a Map algorithm.

Reply:
The linked list would return the link to user space. Not a number.

Removing the link is:

```js
link.next.prev = link.prev;
link.prev.next = link.next;
```

Removing from a Map is O(n) worst case.

The linked list can add more GC work in theory but they'll mostly be young generation.

Reply:
Yeah I think we should use a linked list instead of ids for cancellation, but it's fine to fix that in a follow-up.

Reply:
I like the efficiency. Would be curious to know more about the efficiency of deletion in a Map or object. It sounds like the slowest way is to search in an array and delete with splice, then the next fastest is deleting from a Map, and then even faster is using a linked list.

It's possible that other users of this API are relying on the fact that the 'id' returned is typed as a number, but seems unlikely. I had initially avoided changing the type because it introduces inconsistency in our API for how the 'cancel' method works in our renderer in general.

Reply:
Note that in V8, objects with numeric keys are slow. So if we don't use a Map here (or something else) then it might be a good idea to prefix keys, e.g. `id_${id}`.

Reply:
Thanks - that is useful! Will keep it in mind if we end up keeping the map/array approach.

I have a PR that moves this to a linked list[1], but before landing that needed to open a PR to increase test coverage.[2]

[1]: https://github.com/facebook/react/pull/12844
[2]: https://github.com/facebook/react/pull/12858

Reply:
I need to bookmark this, always took me a while to find this comments.

Even today's scheduler uses minHeap instead of linked list, I think this discussion in the past is super valuable: it well represents how React core team thinks of "optimization" and "performance", not just using a `{}` or `[]` to cache value.

I will probably use this as an argument when I heard "it makes no sense to interview frontend with data structure and algorithm"..well in this context it does make sense :)